### PR TITLE
Pipa

### PR DESCRIPTION
**Pull request by fireworkmarks**
### 1. What does this change do, exactly?
Inserted support for cut graph component Pipa

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I am willing to help maintain this change if there are issues with it later

### CODE REVIEW

File: conf/caddy.sql [hunk 1]
```diff
@@ -25,9 +25,20 @@ CREATE TABLE `custom_domain`  (
   `project_id` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
   `host_domain` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
   `bucket_domain` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
-  `tls_domain` BLOB DEFAULT NULL,
-  `tls_domain_key` BLOB DEFAULT NULL,
+  `tls_domain` blob NULL,
```
Comment (cxt90730) [inline conf/caddy.sql]: 'NOT NULL' or 'DEFAULT NULL'?

File: conf/caddy.sql [hunk 2]
```diff
@@ -25,9 +25,20 @@ CREATE TABLE `custom_domain`  (
...
+  `tls_domain_key` blob NULL,
   PRIMARY KEY (`project_id`, `host_domain`) USING BTREE
-) ENGINE = InnoDB CHARACTER SET = utf8 COLLATE = utf8_general_ci;
+) ENGINE = InnoDB CHARACTER SET = utf8 COLLATE = utf8_general_ci ROW_FORMAT = Compact;
+
+-- ----------------------------
+-- Table structure for pipa
+-- ----------------------------
+DROP TABLE IF EXISTS `pipa`;
```
Comment (cxt90730) [inline conf/caddy.sql]: No body knows what is the 'pipa'... `images_style` is better

File: conf/caddy.sql [hunk 3]
```diff
@@ -4,17 +4,17 @@
...
  Target Server Type    : MySQL
  Target Server Version : 50725
  File Encoding         : 65001
 
- Date: 26/09/2019 16:30:02
+ Date: 11/12/2019 17:42:22
 */
 
-SET NAMES utf8;
+SET NAMES utf8mb4;
```
Comment (cxt90730) [inline conf/caddy.sql]: in TiDB , utf8 is the same as utf8mb4, but not support utf8mb4

File: caddyhttp/pipa/validAndParse.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"encoding/base64"
+	. "github.com/journeymidnight/yig-front-caddy/caddyerrors"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"unicode/utf8"
+)
+
+var fuc = []string{"resize", "watermark"}
```
Comment (cxt90730) [inline caddyhttp/pipa/validAndParse.go]: Use constant, and every words need be transfered to lowercase

File: caddyhttp/pipa/process.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+	Url  string `json:"url"`
+}
+
+type result struct {
+	resByte []byte
+	resErr  error
+}
+
+func processRequest(r *http.Request, key string) (response []byte, err error) {
+	fuc := strings.Split(key, "/")
```
Comment (cxt90730) [inline caddyhttp/pipa/process.go]: imageFunc is better, odd word 'fuc'

File: caddydb/types/imageStyle.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+func (i *ImageStyle) InsertStyle() (string, []interface{}) {
+	sql := "insert into pipa(bucket_name, style_name, style_code) values(?,?,?)"
+	args := []interface{}{i.Bucket, i.StyleName, i.StyleCode}
+	return sql, args
+}
+
+func (i *ImageStyle) UpdateStyle() (string, []interface{}) {
+	sql := "update pipa set style_code=? where bucket_name=? and style_name=?"
```
Comment (cxt90730) [inline caddydb/types/imageStyle.go]: the same

File: caddyhttp/pipa/validAndParse.go (hunk 4 above)
Comment (fireworkmarks) [reply caddyhttp/pipa/validAndParse.go]: There are some doubts, I only called it once here, I don't feel the need to use constants

File: conf/caddy.sql [hunk 7]
```diff
@@ -25,9 +25,21 @@ CREATE TABLE `custom_domain`  (
   `project_id` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
   `host_domain` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
   `bucket_domain` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
-  `tls_domain` BLOB DEFAULT NULL,
-  `tls_domain_key` BLOB DEFAULT NULL,
+  `tls_domain` blob DEFAULT NULL,
+  `tls_domain_key` blob DEFAULT NULL,
   PRIMARY KEY (`project_id`, `host_domain`) USING BTREE
-) ENGINE = InnoDB CHARACTER SET = utf8 COLLATE = utf8_general_ci;
+) ENGINE = InnoDB CHARACTER SET = utf8 COLLATE = utf8_bin ROW_FORMAT = Compact;
```
Comment (cxt90730) [inline conf/caddy.sql]: check it

**Pull request merged**